### DISCUSSION TITLE
brun.cmd

### DISCUSSION THREAD

**Discussion by ghost** [objeck/objeck-lang#127] [General]
I tried to generate the GTK bindings myself. No document about how to use the generator. I tried to run brun.cmd but it ended up with error. code_gen.obe is made but it needs input. Unfortunately, there is no document about how the input looks like. Only this:

```
IF NOT "%~1"=="" IF NOT "%~2"=="" IF NOT "%~3"=="" (
	obr code_gen.obe %1 %2 %3
)
```

Who know what `%~1 %~2 %~3` is.

**Comment by objeck** [comment]
I will write a short how-to tomorrow. The binding generator generates code that is about 80% correct the other 20% needs to be tweaked.

**Reply by ghost** [reply]
Will wait for your short how-to :+1:

**Reply by ghost** [reply]
@objeck

**Reply by objeck** [reply]
Wrapping up web server library updates. I will look at GTK next before mlpack.

**Reply by ghost** [reply]
I only want you to tell me how to use the binding generator. I want to mess with it a bit. I didn't ask you to work on either GTK3 or dwindows.

**Reply by ghost** [reply]
dwindows only has a single header (dw.h) so it's a good starting point if I want to mess with the binding generator. I don't ask you to make a binding for dwindows.

**Reply by objeck** [reply]
Ok. 

The output from the generator are in the ``code_gen/gens`` directory. You may want to check out ``gtkwindow.txt`` for reference. The paradigm is that Objeck code goes into the ``compiler/lib_src/gtk3.obs`` file, and C code goes into the ``gtk3.cpp`` file. 

The input parameters for producing new code gens are: ``file`` path to a header file like ``gtkwindow.h``, ``version`` (such as Gtk ``3`` or ``4``), and ``class_name`` (matching something like GtkWindow). 

Note the code pushed this far was to test the GTK code generator. The overall project needs attention to start the make the code valid.

**Reply by ghost** [reply]
Perhaps the original sdl code generator is more easier to work with as it doesn't have the `version` parameter:

https://github.com/objeck/objeck-lang/tree/master/core/lib/sdl/code_gen

**Reply by objeck** [reply]
Yes, it is; however, it only matches patterns in SDL2 header files.

**Reply by ghost** [reply]
@objeck How to invoke the SDL2 generator? I tried to run `obr code_gen.obe test test.h`, the only thing I could see it printed on the screen is filename='test.h' and done, there is no other outputs. If I give full path to test and test.h, still only the line filename='' printed and nothing else. I'm really confused.

**Reply by ghost** [reply]
Please give me a command that works, not a description. I tried to read your code, but I have no idea why it's this way.

**Reply by objeck** [reply]
Please refer to 234768986.

**Reply by objeck** [reply]
Also, the code generator could be better. It produces stub code that has to be hand tweaked. Most of the changes required can be made in under a minute, but that time added up, given the size of the SDL2 library.

**Comment by objeck** [comment]
I ran into a build issue with GTK3 on Windows using ``vcpkg``; however, I can build GTK4. The current GTK3 bindings do not have the shared libraries required for Windows. This is not an issue for Linux or MSYS2 however is a blocker for cross-platform development. I will look at this later and may switch to experimenting with GTK4.

**Reply by ghost** [reply]
GTK is C so MSVC will have no problems linking with the libraries compiled by MSYS2 (UCRT64). Most GTK based projects out there also use MSYS2 to build for Windows if not cross compiled from Linux.

**Reply by ghost** [reply]
Btw I tried to run the binding generator myself because I wanted to adapt it to generate binding for this library (dw.h and dwib.h):

https://dbsoft.org/newsitem.php?id=23

**Reply by objeck** [reply]
Thanks, I will look more as I wrap up the [web server library](https://github.com/objeck/objeck-web-server) efforts. GTK 3/4 is enormous, and the C headers must be matched up with C implementation files to make code generation feasible. This will allow code documentation to be auto-generated. 

I am also playing around with the ``mlpack`` ML library, which has a general binding framework but needs improvement for Windows, i.e., Unicode support.

**Reply by ghost** [reply]
Did you ever have a look at gobject introspection? It's the preferred way to make bindings for GTK. So far you have only tried to parse the GTK headers directly. Why don't you give gobject introspection a chance?

**Reply by ghost** [reply]
Automate bindings generation from metadata is the trend now. Even Microsoft jumped into it:

https://github.com/microsoft/win32metadata

Manual parsing of headers maybe something of the past.

**Comment by objeck** [comment]
I have not spent much time on GTK3 as I am not interested in desktop GUI programming. 

I enjoy building 2D game programs, thus, the time spent on SDL2. My time is now focused on web programming and APIs. We will likely have GTK3 bindings, as the solution seems algorithmic and can be modeled.

Is this a comminty effort?

**Comment by objeck** [comment]
Made some [progress](https://github.com/objeck/objeck-lang/tree/master/core/lib/experimental/gtk) with GTK. The PoC is functional and code generation is coming along.
